### ISSUE TITLE
Can redirect to an abstract class

### ISSUE THREAD

**Issue by munificent** [dart-lang/sdk#30586]
Analyzer with strong mode currently allows:

```dart
abstract class A {
  factory A() = B;
  int foo();
}

abstract class B implements A {}

main() {
  new A(); // ???
}
```

This seems bad since it would let you get your hands on an instance of an abstract class.

cc @leafpetersen

**Comment by lrhn** [comment]
A redirecting factory constructor must redirect to another constructor. The other constructor may be a factory constructor too, so the problem isn't *just* the class being redirected to is abstract, but also that its constructor is generative - that is, that it's actually trying to create an instance of an abstract class.

I don't think the specification requires a warning in this case at all, so it's all down to the runtime-error for instantiating an abstract class. Dart 2 should probably make it a compile-time error with a phrase like:
> It is a compile-time error if a redirecting factory constructor redirects to a generative constructor of an abstract class.

(Phrasing is similar to what it says about redirecting to itself, but doesn't need to consider the recursive case because the last redirection in a chain will be a compile-time error by itself).

**Comment by munificent** [comment]
SGTM. 👍 

We definitely want it to be a compile-time error if we're going to completely eliminate AbstractClassInstantiationError.

**Comment by lrhn** [comment]
I just found the relevant text in the spec. It actually says that it's not allowed:

>  The resulting constructor call is governed by the same rules as an instance creation expression using \NEW{} (\ref{instanceCreation}).

That includes not being allowed to point to a generative constructor of an abstract class (which I assume is a compile-time error in Dart 2).

**Comment by askeksa-google** [comment]
This was fixed in 6e406c78e8ca87e1f410aa0a0fe5b6006870a21d.
